pysb/simulator/base.py

```diff
@@ -190,81 +179,67 @@ def param_values(self, new_params):
             return
         if isinstance(new_params, dict):
             for k in new_params.keys():
-                if k not in self.model.parameters.keys():
+                if k not in self._model.parameters.keys():
                     raise IndexError("new_params dictionary has unknown "
                                      "parameter name (%s)" % k)
             self._params = new_params
         else:
             # accept vector of parameter values as an argument
-            if len(new_params) != len(self.model.parameters):
+            if len(new_params) != len(self._model.parameters):
                 raise ValueError("new_params must be the same length as "
                                  "model.parameters")
             if isinstance(new_params, np.ndarray):
                 self._params = new_params
             else:
                 self._params = np.array(new_params)
 
-    @classmethod
-    def execute(cls, model, verbose=False, **kwargs):
-        """
-        Convenience method to instantiate a Simulator object, call its run()
-        method, and return a ``yfull`` object of trajectories.
-
-        Parameters
-        ----------
-        model : pysb.Model
-            Model passed to the simulator constructor
-        verbose : bool
-            Print simulator messages to STDOUT
-        **kwargs: dict
-            Keyword arguments for the constructor
-        """
-        simulator = cls(model, verbose=verbose, **kwargs)
-        simulator.run()
-        return simulator.concs_all()
-
     @abstractmethod
     def run(self, tspan=None, param_values=None, initials=None):
         """Run a simulation.
 
-        Returns nothing. Use appropriate methods to retrieve concentrations
-        after simulation.
+        Implementations should return a :class:`.SimulationResult` object.
         """
-        pass
+        return None
 
-    @property
-    def n_sims(self):
-        if self._y is None:
-            return 0
-        else:
-            return len(self._y)
 
-    def _calc_yobs_yexpr(self):
-        """Calculate ``yobs`` and ``yexpr`` based on values of ``y`` obtained
-        in ``run``."""
-        nsims = self.n_sims
-        exprs = self.model.expressions_dynamic()
+class SimulationResult(object):
+    """
+    Contains the results of a simulation, combined with properties and
+    methods to access them.
+    """
+    def __init__(self, simulator, trajectories):
+        self.squeeze = True
+        self.simulator = type(simulator).__name__
+        self.tout = simulator.tout
+        self._y = trajectories
```

Comment:
Should we also check that the appropriate dimension of `trajectories` (the last one, I think) is compatible with the length of `tout`? I think that's invariant across all possible simulators.